ops/model.py

```diff
@@ -797,6 +1001,314 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                '`self.model.get_secret(label=X)`, you should pass `id` instead.'
+                ''.format(id, operation))
+        if not isinstance(id, str):
+            raise TypeError(id)
+
+        id = id.strip()  # trim any whitespace
+
+        # add the prefix if not there already
+        if not id.startswith(cls._secret_id_prefix):
+            id = cls._secret_id_prefix + id
+
+        if not cls._secret_id_re.match(id):
+            raise ValueError("{!r} is not a valid secret id.".format(id))
+
+        return id
+
+    @classmethod
+    def _validate_content(cls, content: Optional[Dict[str, str]]):
+        if not content:
+            # fixme: ATM you can't secret-set without content (e.g. to only update the label),
+            #  but that is a bug in juju.
+            #  when that is fixed, `secret-set [id] --label X` will be valid.
+            raise ModelError('missing secret value or filename')
+
+        invalid_secret_keys = [key for key in content if not cls._secret_key_re.match(key)]
+        invalid_secret_values = [val for val in content.values() if not isinstance(val, str)]  # type: List[Any]  # noqa
+        if invalid_secret_keys or invalid_secret_values:
+            msg = "Secret content invalid."
+            if invalid_secret_keys:
+                msg += '\nInvalid keys: {}. Secret content keys should be ' \
+                       'at least 3 characters long, ' \
+                       'no number to start, no dash to end.'.format(invalid_secret_keys)
+            if invalid_secret_values:
+                bad_types = [type(obj).__name__ for obj in invalid_secret_values]
+                msg += '\nInvalid values: {}. Secret content values should be of ' \
+                       'type str, not {}.'.format(invalid_secret_values, bad_types)
+            raise ValueError(msg)
+
+    def _validate_ownership(self, operation_name: str,
+                            should_own: bool = True):
+        owns = self._is_owned_by_this_unit
+        if should_own and not owns:
+            raise SecretOwnershipError(
+                'Cannot {} secret {!r} which is owned by a different unit'.format(
+                    operation_name, self))
+        if not should_own and owns:
+            raise SecretOwnershipError(
+                'Cannot {} secret {!r} as an owner. Only consumers can do that.'.format(
+                    operation_name, self))
+
+    def _validate_target_relation(self, operation: str,
+                                  target: Union['Relation', Unit, Application],
+                                  relation: Optional['Relation']):
+        if relation is None and not isinstance(target, Relation):
+            raise TypeError(
+                'cannot {} secret {} to a unit or application '
+                'target without specifying a relation as its scope'.format(operation, self))
+        if relation and isinstance(target, Application) and relation.app is not target:
```

Comment:
Target application should come from the "remote application" property of the relation right? So we can never get this error?